### PR TITLE
Prevent okhttp from adding ;charset=utf8 to ContentType Header

### PR DESCRIPTION
**Pull request by nhunzaker**
## Summary

Before this commit, `fetch()` calls append `"charset=utf8"` to the `Content-Type` header on Android (and not on iOS). This is because of an implementation detail in the okhttp library. This means that you can make a call on the JavaScript side like:

```javascript
let body = JSON.stringify({ key: "value" });

let headers = {
  "Content-Type": "application/json"
};

fetch("http://10.0.2.2:3000", { method: "POST", body, headers });
```

However the resulting request appends the utf8 character:

```
POST - 13:34:32:
  content-type: application/json; charset=utf-8
  content-length: 15
  host: 10.0.2.2:3000
  connection: Keep-Alive
  accept-encoding: gzip
  user-agent: okhttp/3.12.1
```

Passing byte array into the RequestBody avoids this, as recommended by a maintainer of okhttp:

https://github.com/square/okhttp/issues/2099#issuecomment-366757161

Related issues: 
https://github.com/facebook/react-native/issues/8237

## Changelog

[Android][fixed] - Prevent fetch() POST requests on Android from appending `charset=utf-8` to `Content-Type` header.

## Test Plan

I setup a test project here: https://github.com/nhunzaker/rn-charset-issue. It includes a simple node server that logs header output from requests. With this change, the server no longer reports the addition of `charset=utf-8` in the `Content-Type` header.

### CODE REVIEW

File: ReactAndroid/src/test/java/com/facebook/react/modules/network/NetworkingModuleTest.java [hunk 1]
```diff
@@ -309,6 +309,55 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
     assertThat(requestHeaders.get("User-Agent")).isEqualTo("React test agent/1.0");
   }
 
+  @Test
+  public void testPostJsonContentTypeHeader() throws Exception {
```
Comment (nhunzaker) [inline ReactAndroid/src/test/java/com/facebook/react/modules/network/NetworkingModuleTest.java]: I copied this from previous tests. It's possible more of the mocks below can be removed.

**Comment by hey99xx** [comment]
This is actually a pretty simple fix for a long going problem for many 👍
Can we still specify utf8 as the charset for the bytes though?

```java
import com.facebook.react.common.StandardCharsets;

body.getBytes(StandardCharsets.UTF_8)
```

For unit tests, you're heavily mocking OkHttp. Was the test actually failing as intended when you undo your fix?

**Comment by nhunzaker** [comment]
@hey99xx good deal, :+1:. I wondered about that, or passing in the character set as defined by `MediaType.parse`.

When I reverse this change, I get the following failure:

```
org.junit.ComparisonFailure: expected:<null> but was:<UTF-8>
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
```

I'd love to actually assert the headers that OkHTTP is building. This was the closest I could manage to get.

**Comment by cpojer** [comment]
@nhunzaker I'm not sure I understand your last comment. Can you confirm that @hey99xx's concern is addressed?

**Comment by nhunzaker** [comment]
Regarding:

> For unit tests, you're heavily mocking OkHttp. Was the test actually failing as intended when you undo your fix?

Yes, I believe I have. While the test heavily mocks OkHttp, the important part to test is [the stringification of the MediaType class](https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/MediaType.java#L151), which is not mocked.

I sent out another commit that I hope simplifies the test and makes that clearer (https://github.com/facebook/react-native/pull/23580/files#diff-0a0a7fd9969f1f5f68d80bf421fcc092R346).

Without my change, this fails, returning `"application/json; charset=utf-8"`

**Comment by cpojer** [comment]
@nhunzaker I mean this part: "Can we still specify utf8 as the charset for the bytes though?" – I assume the answer is yes, just wanna hear the confirmation.

**Comment by nhunzaker** [comment]
@cpojer Oh! Sorry, yep!

This line should do it:
https://github.com/facebook/react-native/pull/23580/files#diff-f8703d50cf5af4c06824e2928dddb500R375

**Comment by cpojer** [comment]
Yes it definitely does matter ☺️

**Pull request closed without merging**